File: archive/polyglot/java/SwarmIntelligence.java
/**
 * SDACS Swarm Intelligence (Java)
 * ================================
 * Particle Swarm Optimization for drone fleet coordination
 * 
 * Features:
 *   - ACO (Ant Colony Optimization)
 *   - PSO (Particle Swarm Optimization)
 *   - Multi-drone task allocation
 *   - Real-time path re-planning
 */

import java.util.*;
import java.util.concurrent.*;

public class SwarmIntelligence {
    
    public static class Position {
        public double x, y, z;
        
        public Position(double x, double y, double z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
        
        public double distanceTo(Position other) {
            double dx = x - other.x;
            double dy = y - other.y;
            double dz = z - other.z;
            return Math.sqrt(dx*dx + dy*dy + dz*dz);
        }
        
        public Position add(Position other) {
            return new Position(x + other.x, y + other.y, z + other.z);
        }
        
        public Position subtract(Position other) {
            return new Position(x - other.x, y - other.y, z - other.z);
        }
        
        public Position multiply(double scalar) {
            return new Position(x * scalar, y * scalar, z * scalar);
        }
        
        @Override
        public String toString() {
            return String.format("(%.2f, %.2f, %.2f)", x, y, z);
        }
    }
    
    public static class Drone {
        public String id;
        public Position position;
        public Position velocity;
        public Position bestPosition;
        public double bestFitness;
        public double fitness;
        
        public Drone(String id, Position start) {
            this.id = id;
            this.position = start;
            this.velocity = new Position(0, 0, 0);
            this.bestPosition = start;
            this.bestFitness = Double.MAX_VALUE;
            this.fitness = Double.MAX_VALUE;
        }
    }
    
    public static class SwarmOptimizer {
        private List<Drone> drones;
        private Position globalBest;
        private double globalBestFitness;
        
        private double inertiaWeight = 0.7;
        private double cognitiveWeight = 1.5;
        private double socialWeight = 1.5;
        
        private Random random;
        
        public SwarmOptimizer(int numDrones) {
            this.random = new Random();
            this.drones = new ArrayList<>();
            this.globalBestFitness = Double.MAX_VALUE;
            
            for (int i = 0; i < numDrones; i++) {
                Position start = new Position(
                    random.nextDouble() * 500,
                    random.nextDouble() * 500,
                    10 + random.nextDouble() * 90
                );
                Drone drone = new Drone("DRONE-" + i, start);
                drone.fitness = evaluateFitness(drone.position);
                drone.bestFitness = drone.fitness;
                drone.bestPosition = drone.position;
                drones.add(drone);
                
                if (drone.fitness < globalBestFitness) {
                    globalBestFitness = drone.fitness;
                    globalBest = drone.position;
                }
            }
        }
        
        private double evaluateFitness(Position pos) {
            Position target = new Position(250, 250, 50);
            double distanceScore = pos.distanceTo(target);
            
            double obstaclePenalty = 0;
            double centerDist = Math.sqrt(
                Math.pow(pos.x - 250, 2) + Math.pow(pos.y - 250, 2)
            );
            if (centerDist < 50) {
                obstaclePenalty = 1000;
            }
            
            double energyCost = pos.z * 0.1;
            
            return distanceScore + obstaclePenalty + energyCost;
        }
        
        private void updateVelocity(Drone drone) {
            double r1 = random.nextDouble();
            double r2 = random.nextDouble();
            
            Position cognitive = drone.bestPosition.subtract(drone.position)
                .multiply(cognitiveWeight * r1);
            Position social = globalBest.subtract(drone.position)
                .multiply(socialWeight * r2);
            Position inertia = drone.velocity.multiply(inertiaWeight);
            
            drone.velocity = inertia.add(cognitive).add(social);
            
            double maxVel = 50.0;
            double velMag = drone.velocity.distanceTo(new Position(0, 0, 0));
            if (velMag > maxVel) {
                drone.velocity = drone.velocity.multiply(maxVel / velMag);
            }
        }
        
        private void updatePosition(Drone drone) {
            drone.position = drone.position.add(drone.velocity);
            
            drone.position.x = Math.max(0, Math.min(500, drone.position.x));
            drone.position.y = Math.max(0, Math.min(500, drone.position.y));
            drone.position.z = Math.max(10, Math.min(100, drone.position.z));
            
            drone.fitness = evaluateFitness(drone.position);
            
            if (drone.fitness < drone.bestFitness) {
                drone.bestFitness = drone.fitness;
                drone.bestPosition = drone.position;
                
                if (drone.fitness < globalBestFitness) {
                    globalBestFitness = drone.fitness;
                    globalBest = drone.position;
                }
            }
        }
        
        public Position optimize(int iterations) {
            for (int iter = 0; iter < iterations; iter++) {
                for (Drone drone : drones) {
                    updateVelocity(drone);
                    updatePosition(drone);
                }
            }
            return globalBest;
        }
        
        public Position getBestPosition() { return globalBest; }
        public double getBestFitness() { return globalBestFitness; }
    }
    
    public static class AntColony {
        private int numAnts;
        private double alpha = 1.0;
        private double beta = 2.0;
        private double evaporation = 0.5;
        private double[][] pheromones;
        private List<Position> targets;
        
        public AntColony(int numAnts, List<Position> targets) {
            this.numAnts = numAnts;
            this.targets = targets;
            this.pheromones = new double[targets.size()][targets.size()];
            
            for (int i = 0; i < targets.size(); i++) {
                for (int j = 0; j < targets.size(); j++) {
                    pheromones[i][j] = 1.0;
                }
            }
        }
        
        public List<Integer> findBestPath() {
            List<Integer> bestPath = new ArrayList<>();
            double bestLength = Double.MAX_VALUE;
            
            for (int ant = 0; ant < numAnts; ant++) {
                List<Integer> path = new ArrayList<>();
                Set<Integer> visited = new HashSet<>();
                int current = random.nextInt(targets.size());
                path.add(current);
                visited.add(current);
                
                while (visited.size() < targets.size()) {
                    int next = selectNext(current, visited);
                    path.add(next);
                    visited.add(next);
                    current = next;
                }
                
                double length = calculatePathLength(path);
                if (length < bestLength) {
                    bestLength = length;
                    bestPath = path;
                }
            }
            
            updatePheromones(bestPath);
            return bestPath;
        }
        
        private int selectNext(int current, Set<Integer> visited) {
            double[] probabilities = new double[targets.size()];
            double sum = 0;
            
            for (int i = 0; i < targets.size(); i++) {
                if (!visited.contains(i)) {
                    double pheromone = Math.pow(pheromones[current][i], alpha);
                    double heuristic = Math.pow(1.0 / targets.get(current).distanceTo(targets.get(i)), beta);
                    probabilities[i] = pheromone * heuristic;
                    sum += probabilities[i];
                }
            }
            
            double r = random.nextDouble() * sum;
            double cumulative = 0;
            for (int i = 0; i < probabilities.length; i++) {
                cumulative += probabilities[i];
                if (cumulative >= r) {
                    return i;
                }
            }
            
            return random.nextInt(targets.size());
        }
        
        private double calculatePathLength(List<Integer> path) {
            double length = 0;
            for (int i = 0; i < path.size() - 1; i++) {
                length += targets.get(path.get(i)).distanceTo(targets.get(path.get(i + 1)));
            }
            return length;
        }
        
        private void updatePheromones(List<Integer> path) {
            for (int i = 0; i < pheromones.length; i++) {
                for (int j = 0; j < pheromones[i].length; j++) {
                    pheromones[i][j] *= (1 - evaporation);
                }
            }
            
            double contribution = 100.0 / calculatePathLength(path);
            for (int i = 0; i < path.size() - 1; i++) {
                pheromones[path.get(i)][path.get(i + 1)] += contribution;
                pheromones[path.get(i + 1)][path.get(i)] += contribution;
            }
        }
        
        private Random random = new Random();
    }
    
    public static void main(String[] args) {
        System.out.println("=== SDACS Swarm Intelligence (Java) ===");
        System.out.println("PSO + ACO for Drone Fleet Coordination");
        
        System.out.println("\n--- Particle Swarm Optimization ---");
        SwarmOptimizer pso = new SwarmOptimizer(50);
        System.out.println("Initial best fitness: " + pso.getBestFitness());
        
        Position optimal = pso.optimize(100);
        System.out.println("Optimized position: " + optimal);
        System.out.println("Best fitness: " + pso.getBestFitness());
        
        System.out.println("\n--- Ant Colony Optimization ---");
        List<Position> targets = Arrays.asList(
            new Position(50, 50, 30),
            new Position(150, 100, 40),
            new Position(250, 200, 50),
            new Position(350, 300, 40),
            new Position(450, 400, 30)
        );
        
        AntColony aco = new AntColony(20, targets);
        List<Integer> bestPath = aco.findBestPath();
        System.out.println("Best path indices: " + bestPath);
        
        System.out.println("\n=== Optimization Complete ===");
    }
}

File: src/cpp/swarm_intelligence.cpp
/**
 * SDACS Swarm Intelligence (C++)
 * ==============================
 * Particle Swarm Optimization (PSO) for drone path planning
 * 
 * Features:
 *   - Particle Swarm Optimization
 *   - Multi-objective path optimization
 *   - Parallel computation with OpenMP
 *   - 3D trajectory optimization
 */

#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <algorithm>
#include <omp.h>

struct Position {
    double x, y, z;
    
    Position operator+(const Position& other) const {
        return {x + other.x, y + other.y, z + other.z};
    }
    
    Position operator-(const Position& other) const {
        return {x - other.x, y - other.y, z - other.z};
    }
    
    Position operator*(double scalar) const {
        return {x * scalar, y * scalar, z * scalar};
    }
    
    double distanceTo(const Position& other) const {
        double dx = x - other.x;
        double dy = y - other.y;
        double dz = z - other.z;
        return std::sqrt(dx*dx + dy*dy + dz*dz);
    }
};

struct Particle {
    Position position;
    Position velocity;
    Position bestPosition;
    double bestFitness;
    double fitness;
    
    Particle(Position start, double fitness) 
        : position(start), velocity({0,0,0}), bestPosition(start), 
          bestFitness(fitness), fitness(fitness) {}
};

class SwarmOptimizer {
private:
    std::vector<Particle> particles;
    Position globalBest;
    double globalBestFitness;
    
    double inertiaWeight;
    double cognitiveWeight;
    double socialWeight;
    
    std::mt19937 rng;
    
public:
    SwarmOptimizer(int numParticles, double inertia = 0.7, double cognitive = 1.5, double social = 1.5)
        : inertiaWeight(inertia), cognitiveWeight(cognitive), socialWeight(social),
          globalBestFitness(std::numeric_limits<double>::max()) {
        
        std::random_device rd;
        rng = std::mt19937(rd());
        
        std::uniform_real_distribution<double> distX(0, 500);
        std::uniform_real_distribution<double> distY(0, 500);
        std::uniform_real_distribution<double> distZ(10, 100);
        
        for (int i = 0; i < numParticles; i++) {
            Position start = {distX(rng), distY(rng), distZ(rng)};
            double fitness = evaluateFitness(start);
            particles.emplace_back(start, fitness);
            
            if (fitness < globalBestFitness) {
                globalBestFitness = fitness;
                globalBest = start;
            }
        }
    }
    
    double evaluateFitness(const Position& pos) {
        double distanceScore = pos.distanceTo({250, 250, 50});
        
        double obstaclePenalty = 0;
        double centerDist = std::sqrt(std::pow(pos.x - 250, 2) + std::pow(pos.y - 250, 2));
        if (centerDist < 50) {
            obstaclePenalty = 1000;
        }
        
        double energyCost = pos.z * 0.1;
        
        return distanceScore + obstaclePenalty + energyCost;
    }
    
    void updateVelocity(Particle& p) {
        std::uniform_real_distribution<double> dist(0, 1);
        
        double r1 = dist(rng);
        double r2 = dist(rng);
        
        Position cognitive = (p.bestPosition - p.position) * cognitiveWeight * r1;
        Position social = (globalBest - p.position) * socialWeight * r2;
        Position inertia = p.velocity * inertiaWeight;
        
        p.velocity = inertia + cognitive + social;
        
        double maxVel = 50.0;
        double velMag = std::sqrt(p.velocity.x*p.velocity.x + 
                                   p.velocity.y*p.velocity.y + 
                                   p.velocity.z*p.velocity.z);
        if (velMag > maxVel) {
            p.velocity = p.velocity * (maxVel / velMag);
        }
    }
    
    void updatePosition(Particle& p) {
        p.position = p.position + p.velocity;
        
        p.position.x = std::max(0.0, std::min(500.0, p.position.x));
        p.position.y = std::max(0.0, std::min(500.0, p.position.y));
        p.position.z = std::max(10.0, std::min(100.0, p.position.z));
        
        p.fitness = evaluateFitness(p.position);
        
        if (p.fitness < p.bestFitness) {
            p.bestFitness = p.fitness;
            p.bestPosition = p.position;
            
            if (p.fitness < globalBestFitness) {
                globalBestFitness = p.fitness;
                globalBest = p.position;
            }
        }
    }
    
    Position optimize(int maxIterations) {
        #pragma omp parallel
        {
            #pragma omp for
            for (int iter = 0; iter < maxIterations; iter++) {
                for (auto& p : particles) {
                    updateVelocity(p);
                    updatePosition(p);
                }
            }
        }
        
        return globalBest;
    }
    
    Position getBestPosition() const { return globalBest; }
    double getBestFitness() const { return globalBestFitness; }
};

class MultiObjectiveSwarm {
public:
    struct Objective {
        double distance;
        double energy;
        double safety;
    };
    
    static Objective evaluatePath(const std::vector<Position>& path) {
        double totalDistance = 0;
        for (size_t i = 1; i < path.size(); i++) {
            totalDistance += path[i-1].distanceTo(path[i]);
        }
        
        double totalEnergy = 0;
        for (const auto& pos : path) {
            totalEnergy += pos.z * 0.1;
        }
        
        double minObstacleDist = std::numeric_limits<double>::max();
        for (const auto& pos : path) {
            double centerDist = std::sqrt(std::pow(pos.x - 250, 2) + std::pow(pos.y - 250, 2));
            if (centerDist < 50) {
                minObstacleDist = std::min(minObstacleDist, 50 - centerDist);
            }
        }
        
        return {totalDistance, totalEnergy, minObstacleDist};
    }
};

int main() {
    std::cout << "=== SDACS Swarm Intelligence (C++) ===" << std::endl;
    std::cout << "Particle Swarm Optimization for Drone Path Planning" << std::endl;
    std::cout << "OpenMP enabled: " << omp_get_max_threads() << " threads" << std::endl;
    
    int numParticles = 50;
    int maxIterations = 100;
    
    SwarmOptimizer swarm(numParticles);
    
    std::cout << "\nInitial best fitness: " << swarm.getBestFitness() << std::endl;
    
    Position optimal = swarm.optimize(maxIterations);
    
    std::cout << "\n=== Optimization Results ===" << std::endl;
    std::cout << "Best position: (" << optimal.x << ", " << optimal.y << ", " << optimal.z << ")" << std::endl;
    std::cout << "Best fitness: " << swarm.getBestFitness() << std::endl;
    
    std::vector<Position> testPath = {
        {0, 0, 30},
        {100, 100, 40},
        {200, 200, 50},
        {250, 250, 50},
        {300, 300, 40},
        {400, 400, 30}
    };
    
    auto objectives = MultiObjectiveSwarm::evaluatePath(testPath);
    std::cout << "\n=== Test Path Evaluation ===" << std::endl;
    std::cout << "Distance: " << objectives.distance << "m" << std::endl;
    std::cout << "Energy: " << objectives.energy << "Wh" << std::endl;
    std::cout << "Safety margin: " << objectives.safety << "m" << std::endl;
    
    return 0;
}
